### PR TITLE
feat(server): Make Argo Server issue own JWE for SSO. Fixes #4027 & #3873

### LINKED ISSUE #4027: SSO logs out quickly
## Summary 

Get logged out unexpectedly when using SSO.

## Diagnostics

What version of Argo Workflows are you running? v2.10.0-rc3

```yaml
curl 'https://argo.ip-dl-mlp-ppd-usw2.iks.a.intuit.com/api/v1/version' \
  -H 'authority: argo.ip-dl-mlp-ppd-usw2.iks.a.intuit.com' \
  -H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.102 Safari/537.36' \
  -H 'dnt: 1' \
  -H 'accept: */*' \
  -H 'sec-fetch-site: same-origin' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-dest: empty' \
  -H 'referer: https://argo.ip-dl-mlp-ppd-usw2.iks.a.intuit.com/login' \
  -H 'accept-language: en-GB,en-US;q=0.9,en;q=0.8' \
  -H 'cookie: ****"' \
  --compressed
```

```
{"code":16,"message":"failed to verify id_token oidc: token is expired (Token Expiry: 2020-09-14 21:37:51 +0000 UTC)"}
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

Fixes #4027 
See #3873 
Depended on by #3525 
Depended on by #3878   

# Changes

* Instead of using a access token, issue our own JWE (encrypted JWT). This places a greater burden of security on Argo Server.
* Encrypt the JWT using AES-GCM (256) and key management algo is RSA-OAEP-SHA256. This is now opaque to the user. More secure.

# Actions

* [x] Verify we're using the adequate encryption. Should we be using ECC rather than RSA? 

image

### CODE REVIEW

**Comment by alexec** [comment]
@RichiCoder1 , @zachaller , @KevinThompsonYCRX

File: server/auth/sso/jwt_test.go [hunk 1]
```diff
@@ -14,27 +14,27 @@ const token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwi
...
 		}
 	})
 	t.Run("Basic", func(t *testing.T) {
-		claimSet, err := ClaimSetFor(&rest.Config{Username: "my-username"})
+		claims, err := ClaimSetFor(&rest.Config{Username: "my-username"})
 		if assert.NoError(t, err) {
-			assert.Empty(t, claimSet.Iss)
-			assert.Equal(t, "my-username", claimSet.Sub)
+			assert.Empty(t, claims.Issuer)
+			assert.Equal(t, "my-username", claims.Subject)
```
Comment (RichiCoder1) [inline server/auth/sso/jwt_test.go:26]: It likely won't matter, but `iss` and `sub` are the standard names for these fields so future tooling might will have opinions about the naming of these claims.

File: server/auth/sso/sso.go [hunk 2]
```diff
@@ -168,44 +210,42 @@ func (s *sso) HandleCallback(w http.ResponseWriter, r *http.Request) {
...
 	}
-	c := &jws.ClaimSet{}
+	c := &jwt.Claims{}
 	if err := idToken.Claims(c); err != nil {
 		w.WriteHeader(401)
 		_, _ = w.Write([]byte(fmt.Sprintf("failed to get claims: %v", err)))
 		return
 	}
-	token, err := zjwt.ZJWT(rawIDToken)
+	argoClaims := &jwt.Claims{Issuer: issuer, Expiry: jwt.NewNumericDate(time.Now().Add(expiry))}
```
Comment (RichiCoder1) [inline server/auth/sso/sso.go:219]: Ditto with `exp`

**Review by RichiCoder1 — APPROVED** [review]
This all seems pretty reasonable. Very minor comment about claim naming. If/when this JWT expires, do it just bounce to the configured SSO provider which refreshes the token?

**Comment by alexec** [comment]
> Sure thing. Short version:
AES-GCM-256 is great.
RSA-OAEP not great. ECC is better.
Expiry of 10h sounds very long, but maybe that’s the use case requirements, and I don’t have enough context to comment.
The longer version:
There is an RFC on JWT best practices - https://tools.ietf.org/html/rfc8725 - or if that’s a bit much there is a summary here - https://curity.io/resources/architect/api-security/jwt-best-practices
You also mentioned the word sessions, and there are some conflicting opinions on using JWT for sessions, but this heavily depends on your use case (and I unfortunately don’t have enough context), but do make sure to read - https://curity.io/resources/architect/api-security/jwt-best-practices and http://cryto.net/~joepie91/blog/2016/06/19/stop-using-jwt-for-sessions-part-2-why-your-solution-doesnt-work/

File: server/auth/sso/sso.go [hunk 3]
```diff
@@ -2,27 +2,36 @@ package sso
...
 	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
-
-	"github.com/argoproj/argo/server/auth/jws"
 )
 
-const Prefix = "Bearer id_token:"
+const (
+	Prefix                              = "Bearer v2:"
+	issuer                              = "argo-server"                // the JWT issuer
+	expiry                              = 10 * time.Hour               // how long JWT are valid for
```
Comment (jessesuen) [inline server/auth/sso/sso.go:28]: Do we want this tunable without a new release being needed (e.g. environment variable)?

**Review by jessesuen — APPROVED** [review]

**Comment by jessesuen** [comment]
Simple and secure. great work.

File: server/auth/sso/sso.go (hunk 3 above)
Comment (alexec) [reply server/auth/sso/sso.go:28]: yes - at some point - but not MVP

**Pull request merged**